ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/pathbased/common/PathBasedAccount.java

```diff
@@ -150,22 +169,56 @@ public void setBalance(final Wei value) {
 
   @Override
   public Bytes getCode() {
-    if (code == null) {
-      code = context.getCode(address, codeHash).orElse(Bytes.EMPTY);
+    return getAnalyzedCode().getBytes();
+  }
+
+  @Override
+  public Code getAnalyzedCode() {
+    final Code cachedCode =
+        Optional.ofNullable(codeCache).map(c -> c.getIfPresent(codeHash)).orElse(null);
+
+    if (code != null) {
+      if (cachedCode == null) {
+        // if code is already set, but not in the cache, put it there
+        Optional.ofNullable(codeCache).ifPresent(c -> c.put(codeHash, code));
+      }
+
+      return code;
+    }
+
+    // cache hit, but code was not set before, set it
+    if (cachedCode != null) {
+      code = cachedCode;
+      return cachedCode;
     }
+
+    // cache miss, code not set: get the code from the disk, set it and put it in the cache
+    // code cannot be empty here, as it would have been set to EMPTY_CODE in the constructor
+    final Bytes byteCode = context.getCode(address, codeHash).orElse(Bytes.EMPTY);
+    code = new CodeV0(byteCode);
```

Comment:
I suggest using this constructor `CodeV0(final Bytes byteCode, final Hash codeHash) `as we already have the hashcode :